mllib/src/main/scala/org/apache/spark/ml/feature/LSH.scala

```diff
@@ -112,7 +112,9 @@ private[ml] abstract class LSHModel[T <: LSHModel[T]]
       numNearestNeighbors: Int,
       singleProbe: Boolean,
       distCol: String): Dataset[_] = {
-    require(numNearestNeighbors > 0, "The number of nearest neighbors cannot be less than 1")
+    val count = dataset.count()
```

Comment:
sorry for late reply.
1, Since `approxNearestNeighbors` is for query with only one key, it is supposed to be called many times in practice. Is `val count = dataset.count()` too expensive between calls? Can it be precomputed somewhere?
2, Do we need `numNearestNeighbors <= count`? refer to scala's and RDD's behavior:
```
scala> Array(1).take(2)
res1: Array[Int] = Array(1)
```

```
scala> val rdd = sc.range(0, 1)
rdd: org.apache.spark.rdd.RDD[Long] = MapPartitionsRDD[1] at range at <console>:24

scala> rdd.count
res0: Long = 1

scala> rdd.take
take   takeAsync   takeOrdered   takeSample

scala> rdd.take(3)
res1: Array[Long] = Array(0)
```

Reply:
Fair point, but it's hardly the most expensive operation here, before or after. I'd expect a caller has to cache the input for multiple calls to make sense anyway. I think the count + approxQuantile still beats sort + take.

I also take the point about take() but relaxing the requirement won't remove the need for count() here. Does it make more sense semantically? You're right it would be more consistent with the original implementation. Hm. I guess I'm neutral on it, but it's a valid question.

Reply:
According to the [doc](https://github.com/apache/spark/blob/56a65b971d4f1b43c37a5664a3a4e2e4fcad1c14/mllib/src/main/scala/org/apache/spark/ml/feature/LSH.scala#L175) 
"@param numNearestNeighbors The maximum number of nearest neighbors."
It imply that the output should not contain exact `numNearestNeighbors` items.

1, if we do not require `numNearestNeighbors <= count`, then if `singleProbe` is true, this `count` job can be avoided.
2, if we found that `numNearestNeighbors <= count`, we can directly return `modelDatasetWithDist`.

Reply:
It implies it _could_ have fewer, yeah. This would be the only case to return fewer, yes, and you'd just return the input. I don't think singleProbe is used (and could be removed). But the count() is still needed below in any event.

I mean, if the caller cares, they can check count() vs number of neighbors before calling it a bunch of times anyway.